parlai/agents/transformer/modules.py

```diff
@@ -310,86 +277,58 @@ class TransformerEncoder(nn.Module):
-    :param bool learn_positional_embeddings: If off, sinusoidal embeddings are
-        used. If on, position embeddings are learned from scratch.
+    :param str reduction_type: Type of reduction at the end of the encoder.
+    :param int n_segments: Number of segments/lang/sentence embeddings.
     :param bool embeddings_scale: Scale embeddings relative to their dimensionality.
         Found useful in fairseq.
-    :param bool reduction: If true, returns the mean vector for the entire encoding
-        sequence.
-    :param int n_positions:
-        Size of the position embeddings matrix.
-    :param int n_segments:
-        Number of segments/lang/sentence embeddings.
-    :param activation:
-        Type of nonlinear activation. Can be relu or gelu.
-    :param variant:
-        Which transformer architecture to use. Could be AIAYN or XLM.
-        Future versions may support things like GPT-2, ...
-    :param output_scaling:
-        Scale the outputs by a given scalar
     """
 
     def __init__(
         self,
-        n_heads,
-        n_layers,
-        embedding_size,
-        ffn_size,
-        vocabulary_size,
-        embedding=None,
-        dropout=0.0,
-        attention_dropout=0.0,
-        relu_dropout=0.0,
-        padding_idx=0,
-        learn_positional_embeddings=False,
-        embeddings_scale=False,
-        reduction_type='mean',
-        n_positions=1024,
-        activation='relu',
-        variant='aiayn',
-        n_segments=0,
-        output_scaling=1.0,
+        opt: Opt,
+        vocabulary_size: int,
+        embedding: Optional[nn.Embedding] = None,
+        padding_idx: int = 0,
+        reduction_type: str = 'mean',
+        n_segments: Optional[int] = None,
+        embeddings_scale: Optional[bool] = None,
     ):
         super(TransformerEncoder, self).__init__()
 
-        self.embedding_size = embedding_size
-        self.ffn_size = ffn_size
-        self.n_layers = n_layers
-        self.n_heads = n_heads
-        self.dim = embedding_size
-        self.embeddings_scale = embeddings_scale
+        self.embedding_size = opt['embedding_size']
+        self.ffn_size = opt['ffn_size']
+        self.n_layers = (
+            opt['n_encoder_layers']
+            if opt.get('n_encoder_layers', -1) > 0
+            else opt['n_layers']
+        )
+        self.n_heads = opt['n_heads']
+        self.dim = self.embedding_size
+        self.embeddings_scale = embeddings_scale or opt['embeddings_scale']
         self.reduction_type = reduction_type
         self.padding_idx = padding_idx
         # this is --dropout, not --relu-dropout or --attention-dropout
-        self.dropout_frac = dropout
+        self.dropout_frac = opt['dropout']
         self.dropout = nn.Dropout(p=self.dropout_frac)
-        self.variant = variant
-        self.n_segments = n_segments
+        self.variant = opt['variant']
+        self.n_segments = n_segments or opt.get('n_segments', 0)
```

Comment:
Btw if we find ourselves needing to manually specify any parameters or override what's in opt, we can just add another default None keyword arg and do this.

Reply:
i suppose im a bit curious how you chose which options to keep in the init, and which you left for opt?